#### 비모수 통계분석

#### 적합도 검정

In [25]:
# run 검정
from statsmodels.sandbox.stats.runs import Runs
import numpy as np

H0(귀무가설)=멤버십 소지 고객과 비소지 고객의 방문은 무작위로 이루어졌다.   
H1(대립가설)=멤버십 소지 고객과 비소지 고객의 방문은 무작위로 이루어지지 않았다.

In [30]:
# 멤버십소지:1, 멤버십 비소지:0
x=[1,0,0,0,0,1,1,1,1,0,0,0,0,1,1,0,0,0,0,0]
x=np.array(x)
x

array([1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0])

In [29]:
Runs(x).runs_test()

(-1.8277470669267506, 0.06758752074917526)

런 검정결과 양측검정 z값은 -1.82, p값이 0.067로 도출됨으로써 유의수준 0.1에서 유의하다.   
따라서, 멤버십 소지 고객과 멤버십 비소지 고객의 매장방문은 무작위적으로 이루어졌다는 귀무가설을 기각한다.   
방문의 이유는 알수 없으나, 멤버십 소지 고객의 방문이 자연스럽제 못했다고 볼 수 있다.

In [31]:
# Kolmogorov-Smirnov 검정 분석
from statsmodels.stats.diagnostic import kstest_normal

H0(귀무가설)= T여행사 고객들의 고객만족도점수는 정규분포를 따른다.   
H1(대립가설)= T여행사 고객들의 고객만족도점수는 정규분포를 따르지 않는다.

In [32]:
x = [88,75,79,84,68,51,70,75,88,90,92,88,63,72,94,80,78,98,81,67,85,87,79,81,85,48,79,86,53,100,87,80,80,32,60,75,62,82,40,57]
x = np.array(x)

In [34]:
kstest_normal(x, dist='norm') # 검정할 분포 : norm(정규), dist(지수)

(0.16508249090030575, 0.007856999983881514)

검정통계량은 0.165, 유의확를은 0.007로 유의수준 0.05보다 작게 나타났다.   
따라서, 귀무가설을 기각하여 고객만족도 점수는 정규분포를 따르지 않는다고 볼 수 있다.   
이결과를 바탕으로, 모수적 통계기법 보다 비모수적 통계기법을 통해 인사이트를 도출해야 한다.

#### 동질성 검정

In [37]:
# wilcoxon 부호-순위 검정
from scipy.stats import wilcoxon

H0(귀무가설)= 기존 물류 알고리즘과 신규 알고리즘간을 통한 평균 배송시간은 차이가 없다.   
H1(대립가설)= 기존 물류 알고리즘과 신규 알고리즘간을 통한 평균 배송시간은 차이가 있다.

In [39]:
x=[10,30,9,21,35,12,17]
y=[8,27,16,25,30,13,11]
wilcoxon(x,y)

WilcoxonResult(statistic=12.0, pvalue=0.8125)

wicoxon 부호-순위 통계량이 12이며, 이에대한 유의확률이 0.735이므로 귀무가설을 기각하지 못한다.   
따라서, 기존 물류 알고리즘과 신규 알고리즘간의 평균 배송시간은 차이가 없다

In [42]:
#Mann-whitney U 검정

H0(귀무가설)= 공장 A와 공장B의 생산효율성은 동일하다.   
H1(대립가설)= 공장 A와 공장B의 생산효율성은 다르다.

In [47]:
from scipy.stats import mannwhitneyu
x = [12, 11, 13, 14, 15]
y = [16, 15, 17, 19, 20]

print(mannwhitneyu(x,y))

MannwhitneyuResult(statistic=0.5, pvalue=0.015970696353780123)


In [56]:
# 생산량 평균 순위 출력
xy = pd.DataFrame(x+y)
xy['생산량 순위'] = xy.rank(ascending=False)
xy['공장이름'] = ['A','A','A','A','A','B','B','B','B','B']
xy.groupby('공장이름').mean()

0  생산량 순위
공장이름              
A     13.0     7.9
B     17.4     3.1

Man-Whitney U 검정결과 U 통계량은 0.5이며, 유의확률은 0.016으로 유의수준 0.05보다 작으므로 귀무가설을 기각한다.   
즉 공장 A와 공장B의 생산효율성은 다르다고 할 수 있다.   
생산량 평균 순위를 보면 공장 A의 효율성이 떨어지는 것을 알 수 있다.

In [57]:
#Kruskal-Wallis H 검정

H0(귀무가설)= 3개의 공장의 철근 생산량은 모두 동일하다.   
H1(대립가설)= 3개의 공장의 철근 생산량은 모두 동일하지는 않다.

In [58]:
from scipy.stats import kruskal 
a = [35,41,45,42,33,36,47,45,31,32,40,44]
b = [40,38,44,48,45,46,42,39,40,41,38,47]
c = [30,34,38,39,40,41,38,37,40,41,39,38] 

In [59]:
kruskal(a,b,c)

KruskalResult(statistic=6.047476974964328, pvalue=0.04861911622342764)

In [62]:
abc=pd.DataFrame(a+b+c)
abc['생산량순위']=abc.rank(ascending=False)
abc['공장이름']=''
abc.loc[0:12,'공장이름'] = 1
abc.loc[12:24,'공장이름'] = 2
abc.loc[24:,'공장이름'] = 3
abc.groupby('공장이름').mean()

0  생산량순위
공장이름                  
1     39.250000   19.0
2     42.333333   13.0
3     37.916667   23.5

Kruskal-Wallis H 검정결과 카이제곱이 6.05이며, 유의확률은 0.0486으로 0.05보다 작아 귀무가설을 기각한다.  
즉, 3개 공장의 철근 생산량은 모두 동일하지 않다.   
공장 간의 생산능력을 확인하면, 3의 평균 생산량이 가장 낮음을 알 수 있다.

In [63]:
# Friedman 검정

H0(귀무가설)= 혜택별 고객의 선호도 평가에 차이는 없다.  
H1(대립가설)= 혜택별 고객의 선호도 평가에 차이가 있다

In [64]:
from scipy.stats import friedmanchisquare 
a = [1,2,1,1,2]
b = [3,3,3,2,1]
c = [2,4,4,4,3]
d = [4,1,2,3,4] 

friedmanchisquare(a,b,c,d)

FriedmanchisquareResult(statistic=6.359999999999999, pvalue=0.09535032301698126)

Friedman 검정결과 카이제곱이 6.36이며, P값이 0.95로 유의수준 0.05보다 크므로 귀무가설을 기각할 수 없다.   
혜택별 고객 선호도의 차이는 뚜렷하게 나타나지 않는다.   

#### 상관성 검정

In [65]:
# Kendall 서열상관 분석

H0(귀무가설) = H 홈쇼핑의 의류브랜드 가격 서열과 판매량 서열은 연관성이 없다.   
H1(대립가설) = H 홈쇼핑의 의류브랜드 가격 서열과 판매량 서열은 연관성이 있다.

In [66]:
from scipy.stats import kendalltau 
x = [5,2,4,1,3]
y = [4,1,3,2,5] 

#. Kendall 검정 분석 결과 출력
kendalltau(x,y)

KendalltauResult(correlation=0.39999999999999997, pvalue=0.48333333333333334)

Kendall 서열상관분석 결과 상관계수는 0.4, p 값은 0.483으로 도출되었고,  
유의수준 보다 크므로 귀무가설을 기각하지 못하여 의류브랜드 가격 서열과 판매량 서열은 연관성이 없다.